distributed/scheduler.py

```diff
@@ -502,17 +508,28 @@ def add_worker(self, stream=None, address=None, keys=(), ncores=None,
             if self.ncores[address] > len(self.processing[address]):
                 self.idle.add(address)
 
-            for key in list(self.unrunnable):
-                r = self.restrictions.get(key, [])
-                if address in r or host in r or name in r:
-                    self.transitions({key: 'released'})
```

Comment:
Ah, I see that you've pulled it out below.  Will take a look.  

Generally some history here is that for a while I pulled everything out into little small functions like this.  All of the indirection soon became very hard to follow.  I tried to keep everything fairly shallow after the refactor.